src/cache/cache_set.py

```diff
@@ -0,0 +1,275 @@
+    Represents a cache line with MESI protocol state tracking and L1 inclusion status.
+    """
+
+    tag: Optional[int] = None
+    mesi_state: MESIState = MESIState.INVALID
+    in_l1: bool = False  # For cache coherency
+
+    def is_invalid(self) -> bool:
+        """Check if line is in Invalid state"""
+        return self.mesi_state == MESIState.INVALID
+
+    def is_shared(self) -> bool:
+        """Check if line is in Shared state"""
+        return self.mesi_state == MESIState.SHARED
+
+    def is_exclusive(self) -> bool:
+        """Check if line is in Exclusive state"""
+        return self.mesi_state == MESIState.EXCLUSIVE
+
+    def is_modified(self) -> bool:
+        """Check if line is in Modified state"""
+        return self.mesi_state == MESIState.MODIFIED
+
+    def set_invalid(self):
+        """Set line to Invalid state"""
+        self.mesi_state = MESIState.INVALID
+
+    def set_shared(self):
+        """Set line to Shared state"""
+        self.mesi_state = MESIState.SHARED
+
+    def set_exclusive(self):
+        """Set line to Exclusive state"""
+        self.mesi_state = MESIState.EXCLUSIVE
+
+    def set_modified(self):
+        """Set line to Modified state"""
+        self.mesi_state = MESIState.MODIFIED
+
+    def get_state_bits(self) -> int:
+        """Returns the 2-bit encoding of the current MESI state"""
+        return self.mesi_state.bits
+
+    def __str__(self) -> str:
+        """String representation of the cache line"""
+        return (
+            f"Tag={hex(self.tag) if self.tag is not None else 'None'}, "
+            f"State={self.mesi_state.name}, "
+            f"In L1={self.in_l1}"
+        )
+
+
+class CacheSetPLRUMESI:
+    """
+    Implements an N-way set associative cache set with MESI protocol and PLRU replacement.
+    Default configuration is 16-way, with supported ways from 1 to 32 (must be power of 2).
+    """
+
+    MAX_WAYS = 32  # Maximum supported ways
+    MIN_WAYS = 1  # Minimum supported ways
+
+    def __init__(self, num_ways: int = 16):
+        # Validate num_ways
+        if not isinstance(num_ways, int):
+            raise TypeError("Number of ways must be an integer")
+        if num_ways < self.MIN_WAYS or num_ways > self.MAX_WAYS:
+            raise ValueError(
+                f"Number of ways must be between {self.MIN_WAYS} and {self.MAX_WAYS}"
+            )
+        if not (num_ways & (num_ways - 1) == 0):
+            raise ValueError("Number of ways must be a power of 2")
+
+        self.num_ways = num_ways
+        self.tree_levels = int(math.log2(num_ways))  # Levels in PLRU tree
+
+        # Initialize cache lines
+        self.ways = [CacheLine() for _ in range(num_ways)]
+
+        # Initialize PLRU state - only need (num_ways - 1) bits for tree
+        self.state = 0
+
+    def __get_parent(self, node: int) -> int:
+        """Helper method for PLRU tree traversal"""
+        return (node - 1) // 2
+
+    def __update_plru(self, way_index: int) -> None:
```

Comment:
I think the tree traversal is all looking good, as far as I can tell. I've walked through some small examples, and it all seems to make sense. The only thing I'd suggest adding would be some additional commenting / documentation, because it can be a little difficult to understand when you first jump in.

One thing specifically that I think could help would be explaining how the binary tree is mapped to bit numbers - that the root is 0 and you increment going from left to right on each subsequent level. The result is that the way indexes are mapped to node indexes starting with way 0 => node index (num_ways - 1), because that's the number of nodes on all the levels before getting to our last level of the way indexes / leaf nodes.

I mean, there's no real substitute for thinking through it, but I think just a little basic commenting regarding how the tree maps to these index values would be helpful for jump-starting our recall when diving back into the code.

Reply:
@nkanderson I updated the documentation in the source code. Let me know if it makes more sense now.

Reply:
Yeah, this looks great, thanks for adding that!